File: Firebase_Bootcamp/app/src/main/java/com/github/romainlogean/sdp_firebase_bootcamp/MainActivity.java
package com.github.romainlogean.sdp_firebase_bootcamp;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button get_button = (Button) findViewById(R.id.button_get);
        Button set_button = (Button) findViewById(R.id.button_set);

        get_button.setOnClickListener(this::get);
        set_button.setOnClickListener(this::set);
    }

    public void get(View view){
        String phone = ((TextView) findViewById(R.id.phone_textview)).getText().toString();
        Database db = DatabaseManager.getDatabase();

        CompletableFuture<String> future = db.get(phone);

        future.thenAccept(new Consumer<String>() {
            @Override
            public void accept(String s) {
                TextView emailView = (TextView) findViewById(R.id.email_textview);
                emailView.setText(s);
            }
        });
    }

    public void set(View view){
        String email = ((TextView) findViewById(R.id.email_textview)).getText().toString();
        String phone = ((TextView) findViewById(R.id.phone_textview)).getText().toString();

        Database db = DatabaseManager.getDatabase();
        db.set(phone,email);
    }

}

File: app/src/main/java/com/github/lgodier/webapibootcamp/MainActivity.kt
package com.github.lgodier.webapibootcamp

import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import androidx.room.Room
import com.github.lgodier.webapibootcamp.cache.Activity
import com.github.lgodier.webapibootcamp.cache.ActivityDAO
import com.github.lgodier.webapibootcamp.cache.AppDatabase
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import java.util.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        val db = Room.databaseBuilder(
            applicationContext,
            AppDatabase::class.java, "database"
        ).build()

        val activityDao = db.activityDAO()

        val button: Button = findViewById(R.id.GetActivityButton);
        button.setOnClickListener {
            try {
                askForActivity(activityDao)
            } catch (e: Exception) {
                findViewById<TextView>(R.id.textView).text = "Something went wrong :( "     }
        }
    }

    private val retrofit: Retrofit = Retrofit.Builder()
        .baseUrl("https://www.boredapi.com/api/")
        .addConverterFactory(GsonConverterFactory.create())
        .build()

    private fun askForActivity(activityDao: ActivityDAO) {

        val boredApi = retrofit.create(BoredApi::class.java)

        boredApi.getActivity().enqueue(object : Callback<BoredActivity> {
            override fun onResponse(call: Call<BoredActivity>, response: Response<BoredActivity>) {
                if (response.isSuccessful) {
                    var activity = response.body()
                    findViewById<TextView>(R.id.textView).text = "ACTIVITY: " + activity!!.activity
                    Thread{
                        activityDao.insertAll(Activity(activityDao.count()+1, activity.activity))
                    }.start()
                }
            }

            override fun onFailure(call: Call<BoredActivity>, t: Throwable) {
                var message = "This data was cached :"
                Thread {
                    val random = Random()
                    println(random.nextInt(activityDao.count()))
                    message +=
                        activityDao.loadById(random.nextInt(activityDao.count())).description
                }.start()
                findViewById<TextView>(R.id.textView).text = message
                }
        })
    }
   }

    data class BoredActivity(val activity: String)

    interface BoredApi {
        @GET("activity")
        fun getActivity(): Call<BoredActivity>
    }